### PR TITLE
Withdraw Shares Tests

### PR DESCRIPTION
**Pull request by sentilesdal**
Gets us close to parity for withdraw share tests.  The short side is still needed and will land next.  The solidity tests have fuzzing, which are not included in these tests, but can/should be added later.

### CODE REVIEW

File: tests/solidity/test_withdraw_shares.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+"""Withdraw shares tests that match those being executed in the solidity repo."""
+import unittest
+
+import numpy
```
Comment (dpaiton) [inline tests/solidity/test_withdraw_shares.py:4]: The convention is to `import numpy as np`, per [here](https://numpy.org/doc/stable/user/absolute_beginners.html#how-to-import-numpy)

File: tests/solidity/test_withdraw_shares.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        self.alice.wallet.update(wallet_deltas)
+
+    def test_redeem_withdraw_shares_fail_insufficient_shares(self):
+        """Should fail to redeem withdraw shares."""
+        market_state = self.hyperdrive.market_state
+        # advance time and let interest accrue
+        self.block_time.set_time(1)
+        # compund interest = p * e ^(rate * time)
+        # we advance by one half year, and the rate is .05 / year
+        accrued = self.contribution * float(numpy.exp(self.target_apr * 0.5))
```
Comment (dpaiton) [inline tests/solidity/test_withdraw_shares.py]: is this advancing half a year? set_time is the same as line 68, but accrued is using 0.5 here and 1 in line 72

File: tests/solidity/test_withdraw_shares.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    def test_redeem_withdraw_shares_short(self):
+        """Should fail to redeem withdraw shares."""
+        market_state = self.hyperdrive.market_state
+
+        # advance time and let interest accrue
+        self.block_time.set_time(1)
+
+        # compund interest = p * e ^(rate * time)
+        # we advance by one year, and the rate is .05 / year
+        accrued = self.contribution * numpy.exp(self.target_apr * 1)
```
Comment (dpaiton) [inline tests/solidity/test_withdraw_shares.py]: you cast np.exp to float above  but not here

File: tests/solidity/test_withdraw_shares.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                days=self.term_length,
+                time_stretch=pricing_model.calc_time_stretch(self.target_apr),
+                normalizing_constant=self.term_length,
+            ),
+            block_time=self.block_time,
+        )
+        _, wallet_deltas = self.hyperdrive.initialize(self.alice.wallet.address, self.contribution, self.target_apr)
+        self.alice.wallet.update(wallet_deltas)
+
+    def test_redeem_withdraw_shares_fail_insufficient_shares(self):
```
Comment (dpaiton) [inline tests/solidity/test_withdraw_shares.py:54]: Can you mark somewhere that this test will be finished in a followup?

Comment (dpaiton) [reply tests/solidity/test_withdraw_shares.py:54]: & same for other tests in the same boat

File: tests/solidity/test_withdraw_shares.py (hunk 2 above)
Comment (sentilesdal) [reply tests/solidity/test_withdraw_shares.py]: fixed, these will get updated in next PR for shorts

File: tests/solidity/test_withdraw_shares.py (hunk 3 above)
Comment (sentilesdal) [reply tests/solidity/test_withdraw_shares.py]: casted here, nice catch

File: elfpy/markets/hyperdrive/hyperdrive_actions.py [hunk 5]
```diff
@@ -385,28 +578,191 @@ def calc_close_long(
...
+    share_adjustment = -(share_proceeds - share_reserves_delta)
+
+    margin_needs_to_be_freed = (
+        market.market_state.total_supply_withdraw_shares > market.market_state.withdraw_shares_ready_to_withdraw
+    )
+    withdraw_pool_deltas = MarketDeltas()
+    if margin_needs_to_be_freed:
+        open_share_price = market.market_state.checkpoints[mint_time].long_share_price
+        # The withdrawal pool has preferential access to the proceeds generated from closing longs.
+        # The LP proceeds when longs are closed are equivalent to the proceeds of short positions.
```
Comment (dpaiton) [inline elfpy/markets/hyperdrive/hyperdrive_actions.py:603]: long position?

File: elfpy/markets/hyperdrive/hyperdrive_actions.py [hunk 6]
```diff
@@ -385,28 +578,191 @@ def calc_close_long(
...
+
+    return updated_share_reserves, updated_bond_reserves
+
+
+def calc_short_interest(
+    bond_amount: float, open_share_price: float, close_share_price: float, share_price: float
+) -> float:
+    """Calculates the interest in shares earned by a short position.
+    The math for the short's interest in shares is given by:
+
```
Comment (dpaiton) [inline elfpy/markets/hyperdrive/hyperdrive_actions.py:695]: Can you put this in a `..math::` block?

File: elfpy/markets/hyperdrive/hyperdrive_actions.py [hunk 7]
```diff
@@ -385,28 +578,191 @@ def calc_close_long(
...
+    bond_reserves: float
+        The current total bonds in reserve
+    share_reserves_delta:
+        The delta that should be applied to share reserves.
+    """
+    if share_reserves_delta == 0:
+        return 0, 0
+
+    updated_share_reserves = share_reserves + share_reserves_delta
+    updated_bond_reserves = bond_reserves * updated_share_reserves / share_reserves
```
Comment (dpaiton) [inline elfpy/markets/hyperdrive/hyperdrive_actions.py:685]: I think this is equal to: `updated_bond_reserves = bond_reserves * (1 + share_reserves_delta)`

But, the bond reserves are in units of base, right? So wouldn't you need to convert share reserves delta to base?

File: elfpy/markets/hyperdrive/hyperdrive_actions.py [hunk 8]
```diff
@@ -385,28 +578,191 @@ def calc_close_long(
...
+
+def calc_short_proceeds(
+    bond_amount: float, share_amount: float, open_share_price: float, close_share_price: float, share_price: float
+) -> float:
+    """Calculates the proceeds in shares of closing a short position.
+
+    This takes into account the trading profits, the interest that was earned by the short, and the
+    amount of margin that was released by closing the short. The math for the short's proceeds in
+    base is given by:
+
```
Comment (dpaiton) [inline elfpy/markets/hyperdrive/hyperdrive_actions.py:733]: put in `..math::` pls

File: elfpy/markets/hyperdrive/hyperdrive_actions.py [hunk 9]
```diff
@@ -385,28 +578,191 @@ def calc_close_long(
...
+    We convert the proceeds to shares by dividing by the current share price. In the event that the
+    interest is negative and outweighs the trading profits and margin released, the short's proceeds
+    are marked to zero.
+
+    Parameters
+    ----------
+    bond_amount: float
+        The amount of bonds underlying the closed short.
+    share_amount:
```
Comment (dpaiton) [inline elfpy/markets/hyperdrive/hyperdrive_actions.py:745]: type hints

File: tests/solidity/test_close_long.py [hunk 10]
```diff
@@ -74,22 +74,26 @@ def verify_close_long(
...
-            msg=(
-                f"{self.hyperdrive.market_state.bond_reserves=} should equal the "
-                f"time adjusted amount: {(market_state_before.bond_reserves + time_remaining * bond_amount)=}."
-            ),
-        )
-        self.assertEqual(  # share reserves
+        self.assertAlmostEqual(  # share reserves
             self.hyperdrive.market_state.share_reserves,
             market_state_before.share_reserves - agent_base_proceeds / market_state_before.share_price,
+            # TODO: see why this delta is not zero.  100 / 50_000_000 might be rounding error of 0.0002%
```
Comment (dpaiton) [inline tests/solidity/test_close_long.py:80]: This seems like a big enough difference to justify opening up an issue instead of just a TODO flag

**Review by dpaiton — APPROVED** [review]
Left a bunch of comments; once those are resolved it is good to go!

File: tests/solidity/test_withdraw_shares.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+
+        # calculate the portion of the fixed interest that bob owes
+        fixed_interest_owed = (bonds_purchased - base_spent) * lp_shares / market_state.lp_total_supply
+
+        # calculate the expectd_withdrawal_proceeds
+        expected_withdrawal_proceeds = contribution_with_interest - fixed_interest_owed
+
+        _, remove_lp_agent_deltas = self.hyperdrive.remove_liquidity(self.bob.wallet, lp_shares)
+        withdrawal_proceeds = remove_lp_agent_deltas.balance.amount
+        self.assertAlmostEqual(withdrawal_proceeds, expected_withdrawal_proceeds)
```
[the remark this answers is not in the document]
Comment (sentilesdal) [reply tests/solidity/test_withdraw_shares.py:123]: ya good idea

File: tests/solidity/test_withdraw_shares.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+        fixed_interest_owed = (
+            (bonds_purchased + bonds_purchased2 - base_spent2 - base_spent) * lp_shares / market_state.lp_total_supply
+        )
+
+        # calculate the expectd_withdrawal_proceeds
+        expected_withdrawal_proceeds = contribution_with_interest - fixed_interest_owed
+
+        _, remove_lp_agent_deltas = self.hyperdrive.remove_liquidity(self.bob.wallet, lp_shares)
+        withdrawal_proceeds = remove_lp_agent_deltas.balance.amount
+        self.assertAlmostEqual(withdrawal_proceeds, expected_withdrawal_proceeds, delta=100)
```
[the remark this answers is not in the document]
Comment (sentilesdal) [reply tests/solidity/test_withdraw_shares.py:178]: separate needs, will be explicit about places though.

File: elfpy/markets/hyperdrive/hyperdrive_actions.py [hunk 13]
```diff
@@ -108,6 +110,179 @@ def check_action(agent_action: MarketAction) -> None:
...
+
+def calculate_long_adjustment(
+    market_state: hyperdrive_market.MarketState,
+    position_duration: time.StretchedTime,
+    market_time: float,
+) -> float:
+    """Calculates an adjustment amount for lp shares"""
+    if market_time > market_state.long_average_maturity_time:
+        return 0
+    # (year_end - year_start) / (normalizing_constant / 365)
```
[the remark this answers is not in the document]
Comment (sentilesdal) [reply elfpy/markets/hyperdrive/hyperdrive_actions.py:154]: no, market_time is the current time, the maturity_time is the end time, so we are still doing a `end - now` to get the time remaining, then we normalize.

File: elfpy/markets/hyperdrive/hyperdrive_actions.py [hunk 14]
```diff
@@ -108,6 +110,179 @@ def check_action(agent_action: MarketAction) -> None:
...
+    )
+    return calculate_lp_allocation_adjustment(
+        market_state.longs_outstanding,
+        market_state.long_base_volume,
+        normalized_time_remaining,
+        market_state.share_price,
+    )
+
+
+def update_weighted_average(  # pylint: disable=too-many-arguments
```
[the remark this answers is not in the document]
Comment (sentilesdal) [reply elfpy/markets/hyperdrive/hyperdrive_actions.py:166]: yeah everything should get tested honestly :P.  I moved this to a more sane location: elfpy/utils/math/update_weighted_average.py

File: elfpy/markets/hyperdrive/hyperdrive_actions.py [hunk 15]
```diff
@@ -108,6 +110,179 @@ def check_action(agent_action: MarketAction) -> None:
...
+    Since we linearly interpolate between the base volume and the bond amount as the time
+    remaining goes from 1 to 0, the base volume can be determined as follows:
+
+        base_amount = t * base_volume + (1 - t) * bond_amount
+                            =>
+        base_volume = (base_amount - (1 - t) * bond_amount) / t
+    """
+    # If the time remaining is 0, the position has already matured and doesn't have an impact on
+    # LP's ability to withdraw. This is a pathological case that should never arise.
+    if normalized_time_remaining == 0:
```
[the remark this answers is not in the document]
Comment (sentilesdal) [reply elfpy/markets/hyperdrive/hyperdrive_actions.py:227]: I think not?  We'd still want an LP to be able to withdraw.

File: elfpy/markets/hyperdrive/hyperdrive_actions.py [hunk 16]
```diff
@@ -108,6 +110,179 @@ def check_action(agent_action: MarketAction) -> None:
...
+    # Calculate the amount of margin that LPs provided on the long position and update the base
+    # volume aggregates.
+    lp_margin = 0.0
+    # Get the total supply of longs in the checkpoint of the longs being closed. If the longs are
+    # closed before maturity, we add the amount of longs being closed since the total supply is
+    # decreased when burning the long tokens.
+    checkpoint_amount = market.market_state[total_supply][checkpoint_time]
+    # maturity_timestamp = checkpoint_time + market.position_duration.days / 365
+
+    # TODO: remove this?  I think it wasn't in here because we aren't burning yet
```
[the remark this answers is not in the document]
Comment (sentilesdal) [reply elfpy/markets/hyperdrive/hyperdrive_actions.py:268]: Yup!  wasn't sure but now I am.

File: elfpy/markets/hyperdrive/hyperdrive_actions.py (hunk 5 above)
Comment (sentilesdal) [reply elfpy/markets/hyperdrive/hyperdrive_actions.py:603]: eh?

File: elfpy/markets/hyperdrive/hyperdrive_actions.py (hunk 7 above)
Comment (sentilesdal) [reply elfpy/markets/hyperdrive/hyperdrive_actions.py:685]: almost, it's equal to `updated_bond_reserves = bond_reserves * (1 + share_reserves_delta/share_reserves)`, which is just a unitless scalar.

File: tests/solidity/test_close_long.py (hunk 10 above)
Comment (sentilesdal) [reply tests/solidity/test_close_long.py:80]: I agree, task added. https://app.clickup.com/t/866a1uvn2

**Pull request merged**